crates/sui-graphql-rpc/src/types/transaction_block/mod.rs

```diff
@@ -599,3 +509,150 @@ impl TryFrom<TransactionBlockEffects> for TransactionBlock {
+    tx_bounds: TxBounds,
+    scan_limit: Option<u64>,
+    checkpoint_viewed_at: u64,
+) {
+    if !conn.has_previous_page {
+        // There are two special scenarios to consider when `has_previous_page` is false:
+        // 1. The cursor is scan-limited - then the corresponding element would not be in results,
+        //    which means we consequently report that there is no preceding page.
+        // 2. `scan_limit` is set, and we get an empty result set - then we report that there is no
+        //    preceding page.
+        // In both scenarios, we want to override the default behavior and set the new `startCursor`
+        // to 1 greater than the given `after` cursor.
+
+        if let Some(after) = page.after() {
+            if (after.is_scan_limited || scan_limit.is_some())
+                && after.tx_sequence_number > tx_bounds.lo
+            {
+                // When dealing with `scan_limit`, the first scanned transaction serves as the
+                // `startCursor`. This will always be 1 greater than the given `after` cursor.
+                let starting_tx_sequence_number = after.tx_sequence_number + 1;
+
+                // It is possible for the first scanned transaction to show up in the result set,
+                // and if so we should flip `is_scan_limited` to false.
+                let is_scan_limited = transactions.first().map_or(true, |tx| {
+                    tx.tx_sequence_number as u64 != starting_tx_sequence_number
+                });
+                let start_cursor = cursor::TransactionBlockCursor {
+                    checkpoint_viewed_at,
+                    tx_sequence_number: starting_tx_sequence_number,
+                    is_scan_limited,
+                };
+
+                conn.has_previous_page = true;
+                conn.start_cursor = Some(Cursor::new(start_cursor).encode_cursor());
+            }
+        }
+    }
+
+    if !conn.has_next_page && scan_limit.is_some() && tx_bounds.scan_has_next_page() {
+        // There are 4 scenarios that will yield `has_next_page=false`:
+        // 1. met `limit`, `paginate_results` doesn't detect more results + more to scan
+        // 2. met `limit`, `paginate_Results` doesn't detect more results + no more to scan
+        // 3. less than `limit`, `paginate_results` doesn't detect more results + more to scan
+        // 4. less than `limit`, `paginate_results` doesn't detect more results + no more to scan
+        // Regardless of the scenario, we can set the `endCursor` to the last transaction scanned.
+
+        let scan_hi = tx_bounds.scan_hi();
+
+        conn.has_next_page = true;
+        conn.end_cursor = Some(
+            Cursor::new(cursor::TransactionBlockCursor {
+                checkpoint_viewed_at,
+                tx_sequence_number: scan_hi,
+                is_scan_limited: transactions
+                    .last()
+                    .map_or(true, |tx| tx.tx_sequence_number as u64 != scan_hi),
```

Comment:
A similar argument to the one about `start_cursor` being scan limited applies here, I think.